### PR TITLE
Refactor the View package

### PR DESCRIPTION
**Pull request by mbabker**
This is a fully refactored View package for Framework 2.0.  Unfortunately, it pretty much completely breaks B/C to 1.0 except for the interface and the base AbstractView class, but I hope this is worth it.

This PR proposes to refactor the package to expose a data setting API and removes the logic for Joomla's own file rendering in support of an injected `RendererInterface` object which is a wrapper around a third party rendering system.  The [babdev/renderer](https://packagist.org/packages/babdev/renderer) package provides the rendering interface (which will be transferred to Joomla if accepted) and provides a common wrapper API around several third party renderers, to include [Twig](http://twig.sensiolabs.org/), Symfony's [Templating Component](http://symfony.com/doc/current/components/templating/introduction.html), and [Plates](http://platesphp.com/engine/).  This should also enable support for additional renderers (such as the CMS's JDocument classes) as desired.

The abstract HTML class is fully removed in favor of a base HTML class which is usable out-of-the-box with an injected renderer.  As well, a base JSON view class is added to (which is a simple wrapper for json_encode on the stored data).

A goal here is to replace the dependency on model classes as has been conventionally accepted in the Joomla CMS in favor of data being injected via the controller which is then rendered when instructed to do so.  The model dependency is not removed though at this point.

Additionally, the rendering interface should enable support for rendering "partial" layouts or additional rendering use cases (in an application I'm building where I use this version of the view package plus the renderer package, I am able to render Symfony forms via their base classes).

### CODE REVIEW

File: src/AbstractView.php [hunk 1]
```diff
@@ -52,4 +87,49 @@ public function escape($output)
...
+	 *
+	 * @param   string  $key  The array key to remove
+	 *
+	 * @return  void
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function removeData($key)
+	{
+		if (isset($this->data[$key]))
```
Comment (Hackwar) [inline src/AbstractView.php]: Why are you testing with isset() here? Shouldn't it be okay to simply "blindly" unset the key?

File: src/AbstractView.php [hunk 2]
```diff
@@ -52,4 +87,49 @@ public function escape($output)
...
+		{
+			unset($this->data[$key]);
+		}
+	}
```
Comment (Hackwar) [inline src/AbstractView.php]: Should this also return itself for chaining?

File: src/AbstractView.php [hunk 3]
```diff
@@ -39,6 +47,33 @@ public function __construct(ModelInterface $model)
...
+	 * @param   mixed   $value  The data value to add
+	 *
+	 * @return  void
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function addData($key, $value)
+	{
+		$this->data[$key] = $value;
+	}
```
Comment (Hackwar) [inline src/AbstractView.php]: Should this return itself for chaining?

File: src/AbstractView.php [hunk 4]
```diff
@@ -39,6 +47,33 @@ public function __construct(ModelInterface $model)
...
+	 * Resets the internal data array
+	 *
+	 * @return  void
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function clearData()
+	{
+		$this->data = array();
+	}
```
Comment (Hackwar) [inline src/AbstractView.php]: Should this return itself for chaining?

File: src/AbstractView.php [hunk 5]
```diff
@@ -52,4 +87,49 @@ public function escape($output)
...
+	 *
+	 * @return  AbstractView  Method allows chaining
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function setData(array $data)
+	{
+		$this->data = $data;
+
+		return $this;
```
Comment (Hackwar) [inline src/AbstractView.php]: Since you are chaining here, why not in the other cases?

File: src/AbstractView.php [hunk 6]
```diff
@@ -52,4 +87,49 @@ public function escape($output)
...
+	 *
+	 * @param   array  $data  The data array.
+	 *
+	 * @return  AbstractView  Method allows chaining
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function setData(array $data)
+	{
+		$this->data = $data;
```
Comment (Hackwar) [inline src/AbstractView.php]: Would you set this or rather do an ```array_merge()``` here?

File: src/BaseHtmlView.php [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+/**
+ * Part of the Joomla Framework View Package
+ *
+ * @copyright  Copyright (C) 2005 - 2013 Open Source Matters, Inc. All rights reserved.
+ * @license    GNU General Public License version 2 or later; see LICENSE
+ */
+
+namespace Joomla\View;
+
+use BabDev\Renderer\RendererInterface;
```
Comment (Hackwar) [inline src/BaseHtmlView.php]: Are you planning on moving that to the Joomla package or are you pushing in your personal code here? :wink: I would like to see us have this in the final form before merging.

File: src/AbstractView.php (hunk 6 above)
Comment (mbabker) [inline src/AbstractView.php]: My thought here was that calling setData would reset the internal data store.  Maybe a mergeData method for that?  Or users could getData, array_merge(), then setData.  I think I'm fine with either.

File: src/BaseHtmlView.php (hunk 7 above)
Comment (mbabker) [inline src/BaseHtmlView.php]: It'll be moved to Joomla's namespace.  I've been using this code in production for some time and didn't want to have to maintain separate copies of it until acceptance.  If we're good with the renderer package, I can get it all transferred over and work on updating dependencies I've created rather quickly.

File: src/AbstractView.php (hunk 1 above)
Comment (mbabker) [inline src/AbstractView.php]: I think I was going with a "better safe than sorry" approach.  Definitely shouldn't be needed.

File: src/AbstractView.php (hunk 6 above)
Comment (Hackwar) [inline src/AbstractView.php]: I'm thinking that we already have clearData(), so someone who really wants to clear the whole view data would go ```$view->clearData()->setData($data);```, but I would expect a setData() method to not wipe the whole array before setting it...

File: src/AbstractView.php (hunk 5 above)
Comment (mbabker) [inline src/AbstractView.php]: Changed it to support chaining across the board.

File: src/AbstractView.php (hunk 6 above)
Comment (mbabker) [inline src/AbstractView.php]: Very well then, changed the behavior to support merging data.

**Comment by Hackwar** [comment]
I like the simple approach here. How do you plan on using this in production? I mean, I'm thinking about Joomla components in the future here. One of the complaints in correspondence with our MVC was that our layouts are actually views and the views behave more like controllers. So I see that we would not implement a seperate view.php file for this solution, but only work with the controller and layout. At least as far as I understand it. That brings up a few questions:
- Why do we call this View then when it is basically a supporting class and the layout is the actual view?
- Where will the logic be that right now is in our view.php files?
- How do we prevent the discussion that apparently our designers are to stupid to work with PHP, so there needs to be minimal PHP logic in the layout?
- If the logic would go into the controller, how should that look and how do we prevent spaghetti-code there? :smile: 

As I told you several times already, I like the decorator pattern and think we should investigate if this should not be one of our primary design patterns for a future Joomla. In the grand scheme of that idea, this code would fit in there pretty good, since it basically removes the view class from the equation of decoratable objects in a component. I have a gut feeling that it only moves the issue from the view class to the layouts and controllers, but in general I like it. :smile:

**Comment by mbabker** [comment]
So you can kind of go about it two ways, depending on preference.  I've got applications where the controllers are injecting all of the view data and the Base view class is handling the actual rendering and I've got applications (the [Framework's site](https://github.com/joomla/framework.joomla.org/tree/master/src/View/Status) being one) where the views are still fetching data via the model then rendering.  So you could go about getting the data into the renderer whichever way suits you.

There is some design here where you don't actually need to implement child view classes (if you're going about it the controller way).  Again, purposefully built to give some flexibility but also reduce the number of lines of code a consumer would need to implement the package.

The discussion of designers needing to know a bit of any language isn't wholly relevant here.  I think that comes down to the application level and however it wants to render data, but the designer will need to be able to work with whatever rendering platform the application uses (so if the CMS consumes this in full and refactors JDocument to implement the RendererInterface, folks would still need to be aware of how to use it; same as if the CMS ever adopted one of the other renderers).

I don't think this refactoring really solves a lot of CMS specific problems at the moment, but does address a higher level Framework issue (the HTML class sucks IMO; the only time I really used it was for https://github.com/mbabker/jd14bos and that feels like bad code) and introduces some architecture that may be useful down the road, as well as forcing a separation of concerns in the app structure.

**Comment by Hackwar** [comment]
Hmm, I would really like to not loose focus of the CMS here. I fear that we might over-engineer this and I don't want to end in situations where we create rather a CMS for developers than a CMS for, pardon, n00bs. And the framework should always aim to be the basis of the CMS itself.

**Comment by mbabker** [comment]
I think this code can help the CMS massively (I just don't know how yet, trying to drop in any of the MVC triad breaks everything because of typehinting and the differences between the namespaced Framework and un-namespaced Platform).  In theory, components really shouldn't need to change too much.  Instead of setting data to class vars and accessing it in layouts via `$this->whatever`, you'd read your data out of the data store through the renderer's API.  It also causes the layout files to not be included in the scope of a view class, and that one really makes you think about app structure and what data you want available in your layouts (a concern that doesn't exist in the CMS today; I can do everything in a layout thanks to the files being in a class scope and things like JFactory).

**Comment by nueckman** [comment]
Hi Michael, i'm still not sure if `View` can be separated from `Model` and `Controller`. Besides that i would like to see `ViewInterface` to be refactored too. Of course, it depends on the definition of "What's a View", but getting and setting data seem to be more appropriate than escaping.

**Comment by Hackwar** [comment]
I don't see that we should provide something that can be dropped into an existing component as replacement for the current JView, etc. I think we should create a parallel MVC structure, so that extensions can use both. (not at the same time of course)

**Comment by mbabker** [comment]
@nueckman - So I left the interface alone for the moment because honestly I'm not sure what the interface should be.  I think in this structure the `escape` method really doesn't fit in anymore though and should probably be dropped completely.  Otherwise, the view package can pretty much stand without a hard coupling to the rest of the MVC triad.  There's some flexibility now with how you build your views (be it injected from controller or an injected model that's used in a subclass).

@Hackwar - Without some other structure changes in the CMS (I think mainly we need to get the un-namespaced Platform MVC classes to extend their Framework counterparts), I can't even test this in practice against the CMS.  However, I don't see any reason this couldn't be used in extensions if you put the effort in to deal with the current incompatibilities (I'm thinking by 3.5 we can have that dealt with).

**Comment by wilsonge** [comment]
If you feel the interface doesn't fit then I recommend that we take this mass b/c break to re factor.

**Comment by wilsonge** [comment]
Separately I'm not convinced by the removeData method. In the majority of cases you're going to have no clue what the key is going to be of your data.

**Comment by nueckman** [comment]
Well, if there is `addData($key, $data)` there should also be `removeData($key)`. The question is how you are using the View. That's why i'm thinking you can't separate the MVC triad.

**Comment by wilsonge** [comment]
Hmm maybe. I was initially thinking people would generally use setData rather than addData. But actually thinking about it a bit more that's probably not the case

**Comment by mbabker** [comment]
I've dropped the escape method from the interface, it just doesn't fit here anymore.

@nueckman wrote:
> The question is how you are using the View. That's why i'm thinking you can't separate the MVC triad.

So at this point the interface and our base implementations are rather unopinionated.  In theory you could create an MVC layer in Symfony applications (a framework that isn't exactly MVC structured) using our model and view interfaces (controller's rather coupled to our application & input packages) as your base objects.  You could continue to use view classes as you comfortably do in the CMS.  You could write your controllers to inject everything into a view and only use it as a proxy to the rendering platform.  I think there is a lot of flexibility in terms of options right now, we aren't exactly telling users how they **HAVE** to use the interfaces (the abstract and base classes in the packages are a little more strict) but really only defining what is expected if you use them.  At this top level I don't see an issue with the lack of opinion in the code.

The one real architectural change it would force if this code were adopted into the CMS for a future major release is the separation between what you should be doing in the MVC triad and what you should be doing in your layouts.  In theory, no longer should implementors (component developers, template designers, etc.) be trying to access the model from their layouts (completely possible with JViewLegacy because of the scope that layout files are included in), accessing a global factory to request any data from any accessible class in the API, or other practices that aren't really considered favorable in today's practices.  This actually gives some front-end developers exactly what they want; a full separation of concern, enforcing all aspects of the final design of a page being left at the template layer and forcing component level APIs to only return data objects.

**Comment by mbabker** [comment]
So, any more thoughts on this one or compelling reasons to not merge?

**Comment by wilsonge** [comment]
You need to fix conflicts first. But sure

**Comment by mbabker** [comment]
Branch updated.  I'll work on moving the renderer stuff over later on.

**Pull request merged**